File: src/main/java/tasks/Tinkoff/TinkoffGeneration/GraphsDfs/taskD.cpp
#include <iostream>
#include <vector>
#include <string>

bool is_tree(int node, int parent, std::vector<bool>& used, const std::vector<std::vector<int>>& graph) {
    used[node] = true;
    for (int nxt : graph[node]) {
        if (nxt == parent)
    	    continue;
        if (used[nxt] || !is_tree(nxt, node, used, graph))
    		return false;
    }
    return true;
}

int main() {
    int n, m;
    std::cin >> n >> m;
    std::vector<std::vector<int>> graph(n);
    for (int i = 0; i < m; ++i) {
      int frm, to;
      std::cin >> frm >> to;
      frm--, to--;

      graph[frm].push_back(to);
      graph[to].push_back(frm);
    }
    std::vector<bool> used(n, false);

    bool res = is_tree(0, -1, used, graph);
    for (bool u : used) {
        if (!u) {
    	    res = false;
    	    break;
        }
    }

    std::cout << (res ? "YES" : "NO") << std::endl;
    return 0;
}

File: src/main/java/tasks/Tinkoff/TinkoffGenerationС/GraphsDfs/taskD.py
def is_tree(node: int, parent: int, used):
    used[node] = True
    for nxt in graph[node]:
        if nxt == parent:
            continue
        if used[nxt] or not is_tree(nxt, node):
            return False
    return True


n, m = map(int, input().split())
graph = [[] for _ in range(n)]

for _ in range(m):
    frm, to = map(int, input().split())
    frm, to = frm - 1, to - 1
    graph[frm].append(to)
    graph[to].append(frm)

used = [False] * n
res = is_tree(0, -1, used) and all(used)
print('YES' if res else 'NO')
